src/compiler/checker.ts

```diff
@@ -28178,7 +28192,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 const target = (flow as FlowReduceLabel).node.target;
                 const saveAntecedents = target.antecedent;
                 target.antecedent = (flow as FlowReduceLabel).node.antecedents;
-                const result = isReachableFlowNodeWorker((flow as FlowReduceLabel).antecedent, /*noCacheCheck*/ false);
+                const result = isReachableFlowNodeWorker((flow as FlowReduceLabel).antecedent, FlowNodeReachableCacheFlags.None);

```

Comment:
as the comment in this block says:
```ts
// Cache is unreliable once we start adjusting labels
```

We can't let this `isReachableFlowNodeWorker` call here to populate the cache as that would create entries based on the **temporarily** `antecedent`(s). Anything computed below this call can't be preserved, at least not without including the id of this `ReduceLabel` flow in the cache key or smth.

So this disables cache writes. As later on that would be used on the same shared flow but with "correct" `antecedent`s (and that's the only result the compiler should cache and trust).

Reply:
cc @ahejlsberg who has been thinking about this sort of thing recently and I believe had ideas on how to simplify the flow node logic here